### ISSUE TITLE
[bug] Bug in Gradient flow with backward(retain_graph=True) and split()

### ISSUE THREAD

**Issue by ThomasDelteil** [apache/incubator-mxnet#14340]
See below code.

I get 4 different values for my gradient, when I should be getting the same (first one)

```
import mxnet as mx
from mxnet import gluon, autograd, nd
import numpy as np
import random

np.random.seed(1)
mx.random.seed(1)
random.seed(1)

np.random.seed(1)
mx.random.seed(1)
random.seed(1)

class SplitNet(gluon.nn.HybridBlock):
    
    def __init__(self, **kwargs):
        super(SplitNet, self).__init__(**kwargs) 
        with self.name_scope():
            self.main = gluon.nn.Dense(10, prefix='main', activation='sigmoid')
            self.classifier = gluon.nn.Dense(2, prefix='classifier')
    
    def hybrid_forward(self, F, x):
        y = self.main(x)
        z1, z2 = F.split(y, 2)
        out1 = self.classifier(z1)
        out2 = self.classifier(z2)
        return out1, out2
        
ctx=mx.gpu(1)

net = SplitNet(prefix='split_net')
net.initialize(ctx=ctx)

loss_fn = gluon.loss.SoftmaxCELoss()
data = mx.nd.ones((1,5),ctx=ctx)
label = mx.nd.zeros((1,1),ctx=ctx)

with autograd.record():
    out1, out2 = net(data)
    l1 = loss_fn(out1, label)
    l2 = loss_fn(out2, label)
    
for i in range(10):
    net.collect_params().zero_grad()
    l2.backward(retain_graph=True)
    print(net.main.weight.grad())

```
[[ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.01405589  0.01405589  0.01405589  0.01405589  0.01405589]
 [ 0.01213417  0.01213417  0.01213417  0.01213417  0.01213417]
 [-0.00799206 -0.00799206 -0.00799206 -0.00799206 -0.00799206]
 [ 0.0014362   0.0014362   0.0014362   0.0014362   0.0014362 ]
 [-0.00098179 -0.00098179 -0.00098179 -0.00098179 -0.00098179]]
<NDArray 10x5 @gpu(1)>

[[-0.24852175 -0.24852175 -0.24852175 -0.24852175 -0.24852175]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]]
<NDArray 10x5 @gpu(1)>

[[-0.24852175 -0.24852175 -0.24852175 -0.24852175 -0.24852175]
 [ 0.01211889  0.01211889  0.01211889  0.01211889  0.01211889]
 [-0.00798559 -0.00798559 -0.00798559 -0.00798559 -0.00798559]
 [ 0.00143664  0.00143664  0.00143664  0.00143664  0.00143664]
 [-0.00098215 -0.00098215 -0.00098215 -0.00098215 -0.00098215]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]]
<NDArray 10x5 @gpu(1)>

[[-0.24852175 -0.24852175 -0.24852175 -0.24852175 -0.24852175]
 [ 0.01211889  0.01211889  0.01211889  0.01211889  0.01211889]
 [-0.00798559 -0.00798559 -0.00798559 -0.00798559 -0.00798559]
 [ 0.00143664  0.00143664  0.00143664  0.00143664  0.00143664]
 [-0.00098215 -0.00098215 -0.00098215 -0.00098215 -0.00098215]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]]
<NDArray 10x5 @gpu(1)>

[[-0.24852175 -0.24852175 -0.24852175 -0.24852175 -0.24852175]
 [ 0.01211889  0.01211889  0.01211889  0.01211889  0.01211889]
 [-0.00798559 -0.00798559 -0.00798559 -0.00798559 -0.00798559]
 [ 0.00143664  0.00143664  0.00143664  0.00143664  0.00143664]
 [-0.00098215 -0.00098215 -0.00098215 -0.00098215 -0.00098215]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]]
<NDArray 10x5 @gpu(1)>

[[-2.4852175e-01 -2.4852175e-01 -2.4852175e-01 -2.4852175e-01
  -2.4852175e-01]
 [ 1.2118888e-02  1.2118888e-02  1.2118888e-02  1.2118888e-02
   1.2118888e-02]
 [-7.9855938e-03 -7.9855938e-03 -7.9855938e-03 -7.9855938e-03
  -7.9855938e-03]
 [ 1.4366427e-03  1.4366427e-03  1.4366427e-03  1.4366427e-03
   1.4366427e-03]
 [-9.8214624e-04 -9.8214624e-04 -9.8214624e-04 -9.8214624e-04
  -9.8214624e-04]
 [ 3.4956147e-03  3.4956147e-03  3.4956147e-03  3.4956147e-03
   3.4956147e-03]
 [ 3.0333332e-03  3.0333332e-03  3.0333332e-03  3.0333332e-03
   3.0333332e-03]
 [-1.9977894e-03 -1.9977894e-03 -1.9977894e-03 -1.9977894e-03
  -1.9977894e-03]
 [ 3.5870986e-04  3.5870986e-04  3.5870986e-04  3.5870986e-04
   3.5870986e-04]
 [-2.4505850e-04 -2.4505850e-04 -2.4505850e-04 -2.4505850e-04
  -2.4505850e-04]]
<NDArray 10x5 @gpu(1)>

[[-0.24852175 -0.24852175 -0.24852175 -0.24852175 -0.24852175]
 [ 0.01211889  0.01211889  0.01211889  0.01211889  0.01211889]
 [-0.00798559 -0.00798559 -0.00798559 -0.00798559 -0.00798559]
 [ 0.00143664  0.00143664  0.00143664  0.00143664  0.00143664]
 [-0.00098215 -0.00098215 -0.00098215 -0.00098215 -0.00098215]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]]
<NDArray 10x5 @gpu(1)>

[[ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.01405589  0.01405589  0.01405589  0.01405589  0.01405589]
 [ 0.01213417  0.01213417  0.01213417  0.01213417  0.01213417]
 [-0.00799206 -0.00799206 -0.00799206 -0.00799206 -0.00799206]
 [ 0.0014362   0.0014362   0.0014362   0.0014362   0.0014362 ]
 [-0.00098179 -0.00098179 -0.00098179 -0.00098179 -0.00098179]]
<NDArray 10x5 @gpu(1)>

[[-0.24852175 -0.24852175 -0.24852175 -0.24852175 -0.24852175]
 [ 0.01211889  0.01211889  0.01211889  0.01211889  0.01211889]
 [-0.00798559 -0.00798559 -0.00798559 -0.00798559 -0.00798559]
 [ 0.00143664  0.00143664  0.00143664  0.00143664  0.00143664]
 [-0.00098215 -0.00098215 -0.00098215 -0.00098215 -0.00098215]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]]
<NDArray 10x5 @gpu(1)>

[[-0.24852175 -0.24852175 -0.24852175 -0.24852175 -0.24852175]
 [ 0.01211889  0.01211889  0.01211889  0.01211889  0.01211889]
 [-0.00798559 -0.00798559 -0.00798559 -0.00798559 -0.00798559]
 [ 0.00143664  0.00143664  0.00143664  0.00143664  0.00143664]
 [-0.00098215 -0.00098215 -0.00098215 -0.00098215 -0.00098215]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.        ]]
<NDArray 10x5 @gpu(1)>
```

**Comment by ThomasDelteil** [comment]
replacing 
```
        z1, z2 = F.split(y, 2)
        out1 = self.classifier(z1)
        out2 = self.classifier(z2)
```
with
```
        out1 = self.classifier(y[:,:5])
        out2 = self.classifier(y[:,5:])
```
works as expected, I'm thinking the bug might come from the split operator. @eric-haibin-lin

**Comment by ThomasDelteil** [comment]
in case you might have an idea @reminisce @piiswrong @sxjscience

**Comment by ThomasDelteil** [comment]
F.split_v2 does not have this bug.

**Comment by apeforest** [comment]
@ThomasDelteil
I ran it on CPU and didn't get core dump. Here is my output:

```
[[1. 1. 1. 1. 1. 0. 0. 0. 0. 0.]]
[[0. 0. 0. 0. 0. 1. 1. 1. 1. 1.]]

[[1. 1. 1. 1. 1. 0. 0. 0. 0. 0.]]
[[0. 0. 0. 0. 0. 1. 1. 1. 1. 1.]]

[[1. 1. 1. 1. 1. 0. 0. 0. 0. 0.]]
[[0. 0. 0. 0. 0. 1. 1. 1. 1. 1.]]

[[1. 1. 1. 1. 1. 0. 0. 0. 0. 0.]]
[[0. 0. 0. 0. 0. 1. 1. 1. 1. 1.]]
```

We also have a unit test to cover split operator with gradient:
tests/python/unittest/test_infer_type.py

```
def test_split()
    data = mx.nd.arange(16, dtype=np.float64).reshape((4, 4))
    data.attach_grad()
    with autograd.record():
        y = mx.nd.split(data, axis=0, num_outputs=2)
    y[0].backward()
    assert data.grad.dtype == np.float64
```

**Comment by ThomasDelteil** [comment]
@apeforest 

The test does not seem enough to actual test that backward is working, I would add a check on the values of the actual gradients.

Are you using a mkl version of mxnet?
Are you running the first example or the second? I get the bug on ubuntu with a GPU build and also on my max on CPU build:

```
> python test.py
/Users/tdelteil/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
[[1. 1. 1. 1. 1. 0. 0. 0. 0. 0.]]
[[ 0.0000000e+00 -4.6566129e-10 -2.9563751e+35 -1.0844662e-19
  -2.9466291e+35  4.6577470e-10 -2.9332078e+35  1.5849492e+29
   7.3657442e+28  1.0597454e-38]]

[[1. 1. 1. 1. 1. 0. 0. 0. 0. 0.]]
[[ 0.00000000e+00 -4.65661287e-10  0.00000000e+00 -4.65661287e-10
   1.16116826e+33  1.78601853e+31  1.49938936e-43  0.00000000e+00
   0.00000000e+00  0.00000000e+00]]

[[1. 1. 1. 1. 1. 0. 0. 0. 0. 0.]]
[[ 0.0000000e+00 -4.6566129e-10 -2.9551057e+35 -2.5249697e-29
   4.2038954e-45  0.0000000e+00 -6.4196971e-16  4.5826664e-41
   0.0000000e+00 -4.6566129e-10]]

python(78676,0x7fffb122c3c0) malloc: *** error for object 0x7fbfa3e432f8: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```

**Comment by wkcn** [comment]
split has a bug, and it always concatenates the in_grad by dim 1.
However, it still has problem when I fix it.

**Comment by wkcn** [comment]
The bug couldn't be reproduced in this test case.
```python
import mxnet as mx
from mxnet.test_utils import assert_almost_equal
import numpy as np

split_list = [1, 2, 3, 4]
siz = np.prod(split_list)
shape = (siz, siz, siz, siz)
for axis in range(len(split_list)):
    for num_outputs in split_list:

        data = mx.nd.arange(np.prod(shape)).reshape(shape)
        data.attach_grad()

        with mx.autograd.record():
            sp = mx.nd.split(data=data, num_outputs=num_outputs, axis=axis)
            if isinstance(sp, mx.nd.NDArray):
                sp = [sp]
            assert len(sp) == num_outputs, (len(sp), num_outputs)
            loss = mx.nd.sum(mx.nd.stack(*[s * mx.nd.arange(s.size).reshape_like(s) * (i + 1) for i, s in enumerate(sp)], axis=axis))
        loss.backward(retain_graph=True)

        part_len = data.shape[axis] // num_outputs
        grad = data.grad
        for i in range(num_outputs):
            index = [slice(None)] * data.ndim
            index[axis] = slice(i * part_len, (i + 1) * part_len)
            index = tuple(index)
            assert_almost_equal(data[index].asnumpy(), sp[i].asnumpy(), atol=0, rtol=0)
            grad_i = grad[index].asnumpy()
            assert_almost_equal(grad_i, np.arange(grad_i.size).reshape(grad_i.shape) * (i + 1), atol=0, rtol=0)
```

**Comment by Vikas89** [comment]
@ThomasDelteil 
I tried running code  on my mac , 

```
import mxnet as mx
from mxnet import gluon, autograd, nd

for i in range(1):
    ctx = mx.cpu()
    data = mx.nd.ones((1,10,), ctx=ctx)
    param = mx.nd.ones((1,10,), ctx=ctx)
    data.attach_grad(grad_req='write')
    param.attach_grad(grad_req='write')
    with autograd.record():
        z = data*param
#        print(z.shape())
        z1, z2 = z.split(2,1)
        #z1, z2 = z[:,:5], z[:,5:]
    z1.backward(retain_graph=True)
    print(data.grad.asnumpy())
    z2.backward(retain_graph=True)
    print(data.grad.asnumpy())
    print()
    break
```
The results I get is : 
```
[[1. 1. 1. 1. 1. 0. 0. 0. 0. 0.]]
[[ 0.0000000e+00  2.0000000e+00  2.5715471e+36 -4.6577453e-10
   1.6986417e-14  3.6433760e-44  4.8807685e+13  6.0397474e+26
   4.3427447e-20  1.4012985e-45]]
```

I guess, the expected output of  print(data.grad.asnumpy()) is : 
`[[0. 0. 0. 0. 0. 1. 1. 1. 1. 1.]] `

Can you confirm my understanding is correct ? 

I also verified that split_v2 is giving correct results ?

**Comment by ThomasDelteil** [comment]
Yes your understanding is correct

**Comment by Vikas89** [comment]
Ok, the root cause for this is we have a caching in nnvm::legacy_op_util, on first backward pass the data will be inititalized, on next backward pass, split backward gets junk data. 
If I defeat  caching here: https://github.com/apache/incubator-mxnet/blame/master/src/nnvm/legacy_op_util.cc#L146
by making bwd_init_ = false , I get expected output. 

Not sure what was the reason to introduce caching and if what is the consequences if we just remove it.
@tqchen : by any chance do you remember and if it is safe to remove caching ?
